File: app/src/main/java/com/mdkashif/alarm/alarm/battery/BatteryLiveData.java
package com.mdkashif.alarm.alarm.battery;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.BatteryManager;

import com.mdkashif.alarm.alarm.battery.pojo.BatteryStats;

import androidx.lifecycle.LiveData;

public class BatteryLiveData extends LiveData<BatteryStats> {

    private Context context;
    int level;
    float temp;
    String status;

    public BatteryLiveData(Context context) {
        this.context = context;
    }

    @Override
    protected void onActive() {
        super.onActive();
        IntentFilter filter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
        context.registerReceiver(mBatInfoReceiver, filter);
    }

    @Override
    protected void onInactive() {
        super.onInactive();
        context.unregisterReceiver(mBatInfoReceiver);
    }

    private BroadcastReceiver mBatInfoReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context ctxt, Intent intent) {

            final int currLevel = intent.getIntExtra(
                    BatteryManager.EXTRA_LEVEL, -1);
            final int maxLevel = intent.getIntExtra(
                    BatteryManager.EXTRA_SCALE, -1);
            level = (int) Math.round((currLevel * 100.0) / maxLevel);

            if(isConnected(context))
                status="Charging";
            else
                status="Discharging";

            temp = ((float) intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE,0)) / 10;

            postValue(new BatteryStats(level,temp,status));

        }

    };

    public boolean isConnected(Context context) {
        Intent intent = context.registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
        int plugged = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1);
        return plugged == BatteryManager.BATTERY_PLUGGED_AC || plugged == BatteryManager.BATTERY_PLUGGED_USB;
    }
}

File: app/src/main/java/com/mdkashif/universalarm/alarm/battery/misc/BatteryLiveData.kt
package com.mdkashif.universalarm.alarm.battery.misc

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import androidx.lifecycle.LiveData
import kotlin.math.roundToInt

class BatteryLiveData(private val context: Context) : LiveData<BatteryStatsPoJo>() {
    private var level = 0
    private var temp = 0F
    private var status = ""
    private var time = ""

    companion object {
        internal var timeToCharge = ""
        internal var timeToDischarge = ""
    }

    private val mBatteryInfoReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctxt: Context, intent: Intent) {
            val batteryStats = BatteryStats(intent)

            val currLevel = batteryStats.level
            val maxLevel = batteryStats.scale
            level = (currLevel * 100.0 / maxLevel).roundToInt()

            when {
                batteryStats.isCharging -> {
                    time = timeToCharge
                    status = "Charging"
                }
                else -> {
                    time = timeToDischarge
                    status = "Discharging"
                }
            }

            temp = batteryStats.temperature.toFloat()

            postValue(BatteryStatsPoJo(level, temp, status, time))
        }
    }

    override fun onActive() {
        super.onActive()
        val filter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
        context.registerReceiver(mBatteryInfoReceiver, filter)
    }

    override fun onInactive() {
        super.onInactive()
        context.unregisterReceiver(mBatteryInfoReceiver)
    }

    class BatteryTimeStats : BatteryTimeService() {
        override fun onChargingTimePublish(hours: Int, mins: Int) {
            //Called while charging and time to charge is being calculated.
            timeToCharge = "${hours}h ${mins}m"
        }

        override fun onCalculatingChargingTime() {
            //hours and mins params indicates remaining time for full charge.
        }

        override fun onDischargeTimePublish(days: Int, hours: Int, mins: Int) {
            //days, hours and mins params indicates remaining time for discharge.
            timeToDischarge = "${days}d ${hours}h ${mins}m"
        }

        override fun onCalculatingDischargingTime() {
            //Called while discharging and time to discharge is being calculated.
        }

        override fun onFullBattery() {
            //Called when device is charging and battery becomes full.
        }
    }
}
